File: rust/src/bin/p0072.rs
// Project Euler: Problem 72

/*
  https://mathproblems123.wordpress.com/2018/05/10/sum-of-the-euler-totient-function/
*/

use std::collections::HashMap;

euler::run_solver!(72);

fn solve() -> String {
    compute(1_000_000).to_string()
}

fn compute(limit: u64) -> u64 {
    debug_assert!(limit > 1);

    let mut cache: HashMap<u64, u64> = HashMap::new();

    sum_phi(limit, &mut cache) - sum_phi(1, &mut cache)
}

fn sum_phi(num: u64, cache: &mut HashMap<u64, u64>) -> u64 {
    if let Some(&val) = cache.get(&num) {
        val
    } else {
        let mut v = num * (num + 1) / 2;
        for m in 2..=num.isqrt() {
            v -= sum_phi(num / m, cache);
        }
        for d in 1..=(num / (num.isqrt() + 1)) {
            v -= ((num / d) - (num / (d + 1))) * sum_phi(d, cache);
        }

        cache.insert(num, v);
        v
    }
}

#[cfg(test)]
mod tests {
    use super::compute;

    #[test]
    fn p0072_8() {
        assert_eq!(compute(8), 21);
    }
    #[test]
    fn p0072_1000000() {
        assert_eq!(compute(1_000_000), 303963552391);
    }
}

File: ocaml/src/p0072.ml
(* Project Euler: Problem 72 *)

(*
   https://mathproblems123.wordpress.com/2018/05/10/sum-of-the-euler-totient-function/

   recursive version
*)

open Core

let rec sumphi n tbl =
  let sum_1 n =
    let rec aux m acc = if m < 2 then acc else aux (pred m) (acc + sumphi (n / m) tbl) in
    aux (Euler.Math.isqrt n) 0
  in
  let sum_2 n =
    let rec aux d acc =
      if d < 1
      then acc
      else aux (pred d) (acc + (((n / d) - (n / (d + 1))) * sumphi d tbl))
    in
    aux (n / (Euler.Math.isqrt n + 1)) 0
  in
  match Hashtbl.find tbl n with
  | Some v -> v
  | None ->
    let v = (n * (n + 1) / 2) - sum_1 n - sum_2 n in
    Hashtbl.set tbl ~key:n ~data:v;
    v
;;

let compute n =
  let memo_tbl = Hashtbl.create (module Int) in
  sumphi n memo_tbl - sumphi 1 memo_tbl
;;

let solve () = compute 1_000_000 |> Int.to_string

(* Test *)

let%test_unit "8" = [%test_eq: int] (compute 8) 21
let%test_unit "1_000_000" = [%test_eq: int] (compute 1_000_000) 303963552391
